ntbs-service/DataMigration/NotificationImportService.cs

```diff
@@ -0,0 +1,126 @@
+                LogInformation($"Validating notification with Id={linkedNotificationId}");
+
+                var validationErrors = GetValidationErrors(notification);
+                
+                if (validationErrors.Count() == 0)
+                {
+                    notificationsToSave.Add(notification);
+                    LogInformation($"No validation errors found");
+                }
+                else
+                {
+                    LogInformation($"{validationErrors.Count()} validation errors found:");
+                    foreach (var validationError in validationErrors)
+                    {
+                        LogInformation(validationError.ErrorMessage);
+                    }
+                }
+            }
+            LogInformation($"Importing {notificationsToSave.Count()} valid notifications");
+            var savedNotificaitons = await AddLinkedNotificationsAsync(notificationsToSave);
+
+
+            LogInformation($"Finished importing notification with Id={notificationId} and linked notifications");
+            return savedNotificaitons;
+        }
+
+        private IEnumerable<ValidationResult> GetValidationErrors(Notification notification) 
+        {
+            var validationsResults = new List<ValidationResult>();
+
+            notification.ShouldValidateFull = true;  
+            notification.PatientDetails.ShouldValidateFull = true;  
+            notification.ClinicalDetails.ShouldValidateFull = true;  
+            notification.TravelDetails.ShouldValidateFull = true;  
+            notification.VisitorDetails.ShouldValidateFull = true;  
+
+            validationsResults.AddRange(ValidateObject(notification));
+            validationsResults.AddRange(ValidateObject(notification.PatientDetails));
+            validationsResults.AddRange(ValidateObject(notification.ClinicalDetails));
+            validationsResults.AddRange(ValidateObject(notification.TravelDetails));
+            validationsResults.AddRange(ValidateObject(notification.VisitorDetails));
+
+            return validationsResults;
+        }
+
+        private IEnumerable<ValidationResult> ValidateObject<T>(T objectToValidate)
+        {
+            var validationContext = new ValidationContext(objectToValidate);
+            var validationsResults = new List<ValidationResult>();  
+
+            Validator.TryValidateObject(objectToValidate, validationContext, validationsResults, true);
+
+            return validationsResults;
+        }
+
+        private void LogInformation(string message)
+        {
+            Log.Information($"NOTIFICATION IMPORT: {message}");
+        }
+
+        private async Task<List<Notification>> AddLinkedNotificationsAsync(List<Notification> notifications)
+        {
+            var group = new NotificationGroup();
+            _context.NotificationGroup.Add(group);
+
+            notifications = notifications.Select(n => {n.GroupId = group.NotificationGroupId; return n;}).ToList();

```

Comment:
This is dangerous, as you're creating a new list, but actually assigning it to the old variable.
I think you want to either be fully immutable, i.e.
```
var notificationsWithGroup = notifications.Select(n => {n.GroupId = group.NotificationGroupId; return n;}).ToList();
```
or use a more appropriate higher-order-function, here probably a `ForEach`:
```
notifications.ForEach(n => n.GroupId = group.NotificationGroupId);
```

here, I think the latter works nicer!